librz/bin/format/elf/elf_ehdr.c

```diff
@@ -0,0 +1,219 @@
+		RZ_LOG_WARN("Failed to read ELF header (e_phnum) in tiny elf mode.\n")
+	}
+
+	return tmp;
+}
+
+static bool is_tiny_elf(ELFOBJ *bin) {
+	return bin->size == 45;
+}
+
+static bool read_ehdr_other_aux(ELFOBJ *bin, ut64 *offset) {
+	return Elf_(rz_bin_elf_read_half)(bin, offset, &bin->ehdr.e_type) &&
+		Elf_(rz_bin_elf_read_half)(bin, offset, &bin->ehdr.e_machine) &&
+		Elf_(rz_bin_elf_read_word)(bin, offset, &bin->ehdr.e_version) &&
+		Elf_(rz_bin_elf_read_addr)(bin, offset, &bin->ehdr.e_entry) &&
+		Elf_(rz_bin_elf_read_off)(bin, offset, &bin->ehdr.e_phoff) &&
+		Elf_(rz_bin_elf_read_off)(bin, offset, &bin->ehdr.e_shoff) &&
+		Elf_(rz_bin_elf_read_word)(bin, offset, &bin->ehdr.e_flags) &&
+		Elf_(rz_bin_elf_read_half)(bin, offset, &bin->ehdr.e_ehsize) &&
+		Elf_(rz_bin_elf_read_half)(bin, offset, &bin->ehdr.e_phentsize);
+}
+
+static bool read_ehdr_other(ELFOBJ *bin) {
+	ut64 offset = EI_NIDENT;
+
+	if (!read_ehdr_other_aux(bin, &offset)) {
+		RZ_LOG_WARN("Failed to read beginning of the ELF header (until e_phnum).\n")
+		return false;
+	}
+
+	if (!Elf_(rz_bin_elf_read_half)(bin, &offset, &bin->ehdr.e_phnum)) {
+		RZ_LOG_WARN("Failed to read ELF header (e_phnum).\n")
+	}
+
+	if (!Elf_(rz_bin_elf_read_half)(bin, &offset, &bin->ehdr.e_shentsize)) {
+		RZ_LOG_WARN("Failed to read ELF header (e_shentsize).\n")
+	}
+
+	if (!Elf_(rz_bin_elf_read_half)(bin, &offset, &bin->ehdr.e_shnum)) {
+		RZ_LOG_WARN("Failed to read ELF header (e_shnum).\n")
+	}
+
+	if (!Elf_(rz_bin_elf_read_half)(bin, &offset, &bin->ehdr.e_shstrndx)) {
+		RZ_LOG_WARN("Failed to read ELF header (e_shstrndx).\n")
+	}
+
+	if (is_tiny_elf(bin)) {
+		RZ_LOG_WARN("The binary seems to be a tiny elf (45 bytes). Reload e_phnum value.\n");
+		bin->ehdr.e_phnum = get_tiny_elf_phnum(bin);
+	}
+
+	return true;
+}
+
+static bool is_valid_elf_ident(ut8 *e_ident) {
+	return !memcmp(e_ident, ELFMAG, SELFMAG) || !memcmp(e_ident, CGCMAG, SCGCMAG);
+}
+
+static bool print_ehdr_aux(ELFOBJ *bin, PrintfCallback cb, ut64 offset, const char *name, char *(*get_value)(ELFOBJ *bin)) {
+	char *tmp = get_value(bin);
+	if (!tmp) {
+		return false;
+	}
+
+	cb("0x%08" PFMT64x "  %-10s  %s\n", offset, name, tmp);
+
+	free(tmp);
+	return true;
+}
+
+RZ_OWN char *Elf_(rz_bin_elf_get_e_ehsize_as_string)(RZ_NONNULL ELFOBJ *bin) {
+	rz_return_val_if_fail(bin, NULL);
+
+	return rz_str_newf("%d", bin->ehdr.e_ehsize);
+}
+
+RZ_OWN char *Elf_(rz_bin_elf_get_e_entry_as_string)(RZ_NONNULL ELFOBJ *bin) {
+	rz_return_val_if_fail(bin, NULL);
+
+	return rz_str_newf("0x%08" PFMT64x, (ut64)bin->ehdr.e_entry);
+}
+
+RZ_OWN char *Elf_(rz_bin_elf_get_e_flags_as_string)(RZ_NONNULL ELFOBJ *bin) {
+	rz_return_val_if_fail(bin, NULL);
+
+	return rz_str_newf("0x%04x", bin->ehdr.e_flags);
+}
+
+RZ_OWN char *Elf_(rz_bin_elf_get_e_indent_as_string)(RZ_NONNULL ELFOBJ *bin) {
+	rz_return_val_if_fail(bin, NULL);
+
+	return rz_str_newf("%02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x %02x",
+		bin->ehdr.e_ident[0],
```

Comment:
Can't we use something like `rz_hex_bin2str()` instead?

Reply:
`rz_hex_bin2str` is insecure imho. There isn't an explicit size variable for out.

Reply:
You can use `rz_hex_bin2strdup()` then

Reply:
The result is less appealing. What do you think?

```
-paddr      name       vaddr      comment
----------------------------------------------------------------------------------
-0x00000000 MAGIC      0x00000000 7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00
+paddr      name       vaddr      comment
+------------------------------------------------------------------
+0x00000000 MAGIC      0x00000000 7f454c46010101000000000000000000
```

Reply:
> You can use `rz_hex_bin2strdup()` then

Yes, there is an example above.

Reply:
You are right. What I suggest then:
- Create a new API function `rz_hex_bin2str_spaced()` or something
- Use it instead.

It might be helpful/useful in other places as well.

Reply:
> You are right. What I suggest then:
> 
>     * Create a new API function `rz_hex_bin2str_spaced()` or something
> 
>     * Use it instead.
> 
> 
> It might be helpful/useful in other places as well.

Can I bundle this with the `readelf -h` PR?

Reply:
Yes.